snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java

```diff
@@ -0,0 +1,639 @@
+     * @param step         the mean distance from one pixel position to the other in the returned array
+     * @return the rectangular boundary
+     */
+    public static PixelPos[] createPixelBoundary(int rasterWidth, int rasterHeight, Rectangle rect, int step) {
+        if (rect == null) {
+            rect = new Rectangle(0,
+                    0,
+                    rasterWidth,
+                    rasterHeight);
+        }
+        return createRectBoundary(rect, step);
+    }
+
+    /**
+     * Creates a rectangular boundary expressed in pixel positions for the given source rectangle. If the source
+     * {@code rect} is 100 x 50 pixels and {@code step} is 10 the returned array will countain exactly 2 * 10
+     * + 2 * (5 - 2) = 26 pixel positions.
+     * <p>This method is used for an intermediate step when determining a product boundary expressed in geographical
+     * co-ordinates.
+     * <p> This method delegates to {@link #createRectBoundary(java.awt.Rectangle, int, boolean) createRectBoundary(Rectangle, int, boolean)}
+     * and the additional boolean parameter {@code usePixelCenter} is {@code true}.
+     *
+     * @param rect the source rectangle
+     * @param step the mean distance from one pixel position to the other in the returned array
+     * @return the rectangular boundary
+     */
+    public static PixelPos[] createRectBoundary(Rectangle rect, int step) {
+        final boolean usePixelCenter = true;
+        return createRectBoundary(rect, step, usePixelCenter);
+    }
+
+    private static GeoPos[] getGeoBoundary(GeoCoding gc, Rectangle region, int step, boolean usePixelCenter) {
+        final PixelPos[] points = createRectBoundary(region, step, usePixelCenter);
+        ArrayList<GeoPos> geoPoints = new ArrayList<>(points.length);
+        boolean calculateInsets = false;
+        for (final PixelPos pixelPos : points) {
+            final GeoPos gcGeoPos = gc.getGeoPos(pixelPos, null);
+            if (!gcGeoPos.isValid()) {
+                calculateInsets = true;
+                break;
+            }
+            geoPoints.add(gcGeoPos);
+        }
+
+        if (calculateInsets) {
+            geoPoints.clear();
+            geoPoints = createInsetsGeoBoundary(gc, step, region, usePixelCenter);
+        }
+
+        return geoPoints.toArray(new GeoPos[0]);
+    }
+
+    private static ArrayList<GeoPos> createInsetsGeoBoundary(GeoCoding gc, int step, Rectangle rect, boolean usePixelCenter) {
+        final double insetDistance = usePixelCenter ? 0.5 : 0.0;
+        int xStart = rect.x;
+        int yStart = rect.y;
+        final int w = usePixelCenter ? rect.width - 1 : rect.width;
+        final int h = usePixelCenter ? rect.height - 1 : rect.height;
+        int xEnd = xStart + w;
+        int yEnd = yStart + h;
+
+        final ArrayList<GeoPos> geoPosList = new ArrayList<>();
+
+        final int validYMin = getFirstValid_Y(gc, insetDistance, xStart, yStart, xEnd, yEnd);
```

Comment:
Could it make sense to parallelise the computation? Maybe using an ExecutorService?

Reply:
Absolutely